mesonbuild/interpreterbase.py

```diff
@@ -572,68 +590,70 @@ def evaluate_comparison(self, node):
                 'Values of different types ({}, {}) cannot be compared using {}.'.format(type(val1).__name__,
                                                                                          type(val2).__name__,
                                                                                          node.ctype))
-        elif not self.is_elementary_type(val1):
-            raise InterpreterException('{} can only be compared for equality.'.format(node.left.value))
-        elif not self.is_elementary_type(val2):
-            raise InterpreterException('{} can only be compared for equality.'.format(node.right.value))
+        elif not isinstance(val1, self.elementary_types):
+            raise InterpreterException('{} can only be compared for equality.'.format(getattr(node.left, 'value', '<ERROR>')))
+        elif not isinstance(val2, self.elementary_types):
+            raise InterpreterException('{} can only be compared for equality.'.format(getattr(node.right, 'value', '<ERROR>')))
+        # Use type: ignore because mypy will complain that we are comparing two Unions,
+        # but we actually guarantee earlier that both types are the same
```

Comment:
You can use asserts and isinstance to make mypy happy in this case.

Reply:
I have tried `assert type(val1) == type(val2)` but it did not work. The only other solution would be to produce an assert monstrosity like this:

```python
assert (isinstance(val1, str) and isinstance(val2, str)) or
       (isinstance(val1, int) and isinstance(val2, int)) or
       # I think you get the point
```